### PR TITLE
[WIP] Fix: amend commit failure for first commit

### PR DESCRIPTION
**Pull request by asfaltboy**
As described in #557 and  #679 and reproduced in failing test (last commit 1ca6dd3)

### CODE REVIEW

File: tests/test_core_commands/test_gs_commit.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    def test_add_and_commit_file(self):
+        """ Test adding a new file and committing """
+        foo = os.path.join(self.repo_path, "foo")
+        with open(foo, "w") as f:
+            f.write("foo")
+
+        self.stage_file(foo)
+        self.window.run_command("gs_commit")
+
+        yield from self.assert_commit_view_rendered(first_line=None)
```
Comment (stoivo) [inline tests/test_core_commands/test_gs_commit.py]: Can you explain what `yield` yields back too?

Comment (asfaltboy) [inline tests/test_core_commands/test_gs_commit.py]: This specific `yield from` calls into every yield command in the `self.assert_commit_view_rendered` method. Without this `yield from`, simply calling the method would not call the yield expression inside the method, thus failing the test.

For more information see here [the changelog description in Python 3.3][1].

The reason we use yield at all here, is due to the commands' use of `sublime.set_timeout_async` which basically defers and returns control back to the text thread. This a feature of @randy3k 's [excellent UnitTesting plugin][2].

[1]: https://docs.python.org/3/whatsnew/3.3.html#pep-380-syntax-for-delegating-to-a-subgenerator
[2]: https://github.com/randy3k/UnitTesting-example/tree/090f271592788d584d41b92e810b0c7577e066e9

Comment (stoivo) [inline tests/test_core_commands/test_gs_commit.py]: Ok

**Comment by asfaltboy** [comment]
Hey all, I've rebased over latest `dev` branch and added a relatively simple fix that uses log instead of diff to get the content of last commit ... However, it's not realistic diff, as it excludes the staged changes. Does anyone know of a way to get the diff of the "zero-eth" commit?

TODOs:

- [ ] need to improve diff display (either hide it or learn to include staged
      changes) - google it!
- [ ] we also need to test for all the flags we allow to supply to commit w/ amend
- [x] the failing test passes now :)

**Comment by kaste** [comment]
Hi @asfaltboy! 

You should be optimistic here. Like nobody amends its first commit. So try with what we usually do and fallback and repair only if the git command fails.

The parent of the inital commit is `4b825dc642cb6eb9a060e54bf8d69288fbee4904`. So you can use 

```
git diff 4b825dc642cb6eb9a060e54bf8d69288fbee4904 
```

for include unstaged and 

```
git diff 4b825dc642cb6eb9a060e54bf8d69288fbee4904  --cached
```

otherwise if iirc the options we have here.

**Comment by asfaltboy** [comment]
@kaste how do you get the parent of an initial commit? How is that even possible, is there a way to get that 0-state commit?

**Comment by kaste** [comment]
I have it from GitGUI source: https://github.com/git/git/blob/master/git-gui/git-gui.sh#L1423-L1434

This sha is currently a constant. It is though safer to do what gitGUI does. In the console 

```
>>> git = GitCommand()
>>> git.view = view
>>> git.git('mktree', stdin='')
'4b825dc642cb6eb9a060e54bf8d69288fbee4904\n'
```

Look at the empty check -> https://github.com/git/git/blob/master/git-gui/git-gui.sh#L1431

**Comment by kaste** [comment]
Basically, a factoring like this would be dope

```
try:
	diff = self.do_diff(..., parent='HEAD^')
except ...:
	parent = self.git('mktree', stdin='').strip()
	diff = self.do_diff(..., parent=parent)
```

**Comment by asfaltboy** [comment]
That is awesome! I was missing that an empty tree is a hash constant, cool stuff!

Refs:
- https://f-o.org.uk/2017/empty-trees-in-git.html
- https://stackoverflow.com/a/9766506/484127

I will try to play with this in the weekend

**Pull request closed without merging**